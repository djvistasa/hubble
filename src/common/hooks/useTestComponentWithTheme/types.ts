type IUseTestComponentWithTheme = {};

export type { IUseTestComponentWithTheme };
