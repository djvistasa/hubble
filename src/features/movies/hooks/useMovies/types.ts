import { UseMutationResult } from "react-query";

type IUseMoviesReturnValues = {
  getNextPage: () => void;
  searchForMovie: UseMutationResult<
    never[] | undefined,
    unknown,
    string,
    unknown
  >;
  getMovies: () => void;
};

type IMoviesProps = {};

type IMoviesRouteParams = {
  page: string;
};

type ISearchParams = {
  total_pages: string;
};

export type {
  IMoviesProps,
  IMoviesRouteParams,
  ISearchParams,
  IUseMoviesReturnValues,
};
