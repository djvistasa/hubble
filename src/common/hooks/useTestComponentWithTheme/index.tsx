import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import { theme } from "../../theme";

import { RenderResult } from "@testing-library/react";

import { ReactElement } from "react";
import { QueryClient, QueryClientProvider } from "react-query";

function useTestComponentWithTheme() {
  const renderWithTheme = (component: ReactElement<any, any>): RenderResult => {
    const queryClient = new QueryClient();
    return render(
      <QueryClientProvider client={queryClient}>
        <ThemeProvider theme={theme}>{component}</ThemeProvider>
      </QueryClientProvider>
    );
  };

  return renderWithTheme;
}

export default useTestComponentWithTheme;
