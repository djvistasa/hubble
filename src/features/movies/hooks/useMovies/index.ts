import { IMovieResponse } from "@features/movies/types";
import { makeApiRequest } from "@utils";
import { useQuery } from "react-query";
import { useNavigate, useParams, useSearchParams } from "react-router-dom";
import { IMoviesRouteParams, IUseMovies } from "./types";

function useMovies(): IUseMovies {
  // create a getMovies mutation

  const { page } = useParams<IMoviesRouteParams>();
  const [searchParams, setSearchParams] = useSearchParams();

  const navigation = useNavigate();

  console.log(searchParams);

  const getNextPage = () => {
    if (page) {
      const activePage = parseInt(page);
      navigation(`/movies/${activePage + 1}`);
    }
  };

  const { data: movies } = useQuery(["movies"], async () => {
    const { error, result, ok } = await makeApiRequest(
      `/discover/movie?include_adult=false&include_video=false&language=en-US&page=${page}&sort_by=popularity.desc`
    );
    if (error) {
      // handle error
      return [];
    }

    if (result && ok) {
      const { results, total_pages } = result as IMovieResponse;

      setSearchParams({
        total_pages: total_pages.toString(),
      });

      return results;
    }
  });

  return { movies, getNextPage };
}

export default useMovies;
