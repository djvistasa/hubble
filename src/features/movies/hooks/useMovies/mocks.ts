const moviesResp = {
  page: 1,
  results: [
    {
      adult: false,
      backdrop_path: "/pWsD91G2R1Da3AKM3ymr3UoIfRb.jpg",
      genre_ids: [878, 28, 18],
      id: 933131,
      original_language: "ko",
      original_title: "황야",
      overview:
        "After a deadly earthquake turns Seoul into a lawless badland, a fearless huntsman springs into action to rescue a teenager abducted by a mad doctor.",
      popularity: 1907.07,
      poster_path: "/zVMyvNowgbsBAL6O6esWfRpAcOb.jpg",
      release_date: "2024-01-26",
      title: "Badland Hunters",
      video: false,
      vote_average: 6.7,
      vote_count: 406,
    },
    {
      adult: false,
      backdrop_path: "/unvtbkgxh47BewQ8pENvdOdme0r.jpg",
      genre_ids: [28, 18],
      id: 1212073,
      original_language: "de",
      original_title: "60 Minuten",
      overview:
        "Desperate to keep custody of his daughter, a mixed martial arts fighter abandons a big match and races across Berlin to attend her birthday party.",
      popularity: 1212.352,
      poster_path: "/aajCqg315CoJPu1NmgPCkbRjnl6.jpg",
      release_date: "2024-01-19",
      title: "Sixty Minutes",
      video: false,
      vote_average: 6.866,
      vote_count: 351,
    },
    {
      adult: false,
      backdrop_path: "/yyFc8Iclt2jxPmLztbP617xXllT.jpg",
      genre_ids: [35, 10751, 14],
      id: 787699,
      original_language: "en",
      original_title: "Wonka",
      overview:
        "Willy Wonka – chock-full of ideas and determined to change the world one delectable bite at a time – is proof that the best things in life begin with a dream, and if you’re lucky enough to meet Willy Wonka, anything is possible.",
      popularity: 1060.652,
      poster_path: "/qhb1qOilapbapxWQn9jtRCMwXJF.jpg",
      release_date: "2023-12-06",
      title: "Wonka",
      video: false,
      vote_average: 7.198,
      vote_count: 2127,
    },
    {
      adult: false,
      backdrop_path: "/criPrxkTggCra1jch49jsiSeXo1.jpg",
      genre_ids: [878, 12, 28],
      id: 609681,
      original_language: "en",
      original_title: "The Marvels",
      overview:
        "Carol Danvers, aka Captain Marvel, has reclaimed her identity from the tyrannical Kree and taken revenge on the Supreme Intelligence. But unintended consequences see Carol shouldering the burden of a destabilized universe. When her duties send her to an anomalous wormhole linked to a Kree revolutionary, her powers become entangled with that of Jersey City super-fan Kamala Khan, aka Ms. Marvel, and Carol’s estranged niece, now S.A.B.E.R. astronaut Captain Monica Rambeau. Together, this unlikely trio must team up and learn to work in concert to save the universe.",
      popularity: 1039.405,
      poster_path: "/9GBhzXMFjgcZ3FdR9w3bUMMTps5.jpg",
      release_date: "2023-11-08",
      title: "The Marvels",
      video: false,
      vote_average: 6.277,
      vote_count: 1704,
    },
    {
      adult: false,
      backdrop_path: "/ay0PJQZizDXk0pzhoGX4v7K9h7A.jpg",
      genre_ids: [28, 53],
      id: 1214314,
      original_language: "en",
      original_title: "One More Shot",
      overview:
        "Following the attack on the black site in Poland, Navy SEAL Jake Harris is ordered to escort terrorist suspect Amin Mansur to Washington D.C. for interrogation. Before the prisoner transfer process is complete, though, the airport is attacked by a group of heavily armed, well-trained mercenaries.",
      popularity: 989.221,
      poster_path: "/nQ1BQg4yMdlYSHvHZgwladzy7EF.jpg",
      release_date: "2024-01-12",
      title: "One More Shot",
      video: false,
      vote_average: 6.674,
      vote_count: 144,
    },
    {
      adult: false,
      backdrop_path: "/r9oTasGQofvkQY5vlUXglneF64Z.jpg",
      genre_ids: [28, 35],
      id: 1029575,
      original_language: "en",
      original_title: "The Family Plan",
      overview:
        "Dan Morgan is many things: a devoted husband, a loving father, a celebrated car salesman. He's also a former assassin. And when his past catches up to his present, he's forced to take his unsuspecting family on a road trip unlike any other.",
      popularity: 985.331,
      poster_path: "/a6syn9qcU4a54Lmi3JoIr1XvhFU.jpg",
      release_date: "2023-12-14",
      title: "The Family Plan",
      video: false,
      vote_average: 7.331,
      vote_count: 922,
    },
    {
      adult: false,
      backdrop_path: "/w12jjv5XNNpHgM4he1COnfWiA93.jpg",
      genre_ids: [28, 27, 35, 53],
      id: 1211483,
      original_language: "en",
      original_title: "Skal - Fight for Survival",
      overview:
        "My name's Arthur, a huge Internet star who's just hit 3 million subs. While in the midst of throwing an epic party to celebrate, the universe had the balls to bring on the effing apocalypse and cut my night short. What was supposed to be a perfect hangover, has turned into an epic fight for survival.",
      popularity: 967.767,
      poster_path: "/uQkiDKQyun13mqsOXv7I5MRKr0q.jpg",
      release_date: "2023-11-24",
      title: "Skal - Fight for Survival",
      video: false,
      vote_average: 5.913,
      vote_count: 23,
    },
    {
      adult: false,
      backdrop_path: "/rz8GGX5Id2hCW1KzAIY4xwbQw1w.jpg",
      genre_ids: [28, 35, 80],
      id: 955916,
      original_language: "en",
      original_title: "Lift",
      overview:
        "An international heist crew, led by Cyrus Whitaker, race to lift $500 million in gold from a passenger plane at 40,000 feet.",
      popularity: 921.295,
      poster_path: "/46sp1Z9b2PPTgCMyA87g9aTLUXi.jpg",
      release_date: "2024-01-10",
      title: "Lift",
      video: false,
      vote_average: 6.586,
      vote_count: 791,
    },
    {
      adult: false,
      backdrop_path: "/uUiIGztTrfDhPdAFJpr6m4UBMAd.jpg",
      genre_ids: [878, 28, 12],
      id: 634492,
      original_language: "en",
      original_title: "Madame Web",
      overview:
        "Forced to confront revelations about her past, paramedic Cassandra Webb forges a relationship with three young women destined for powerful futures...if they can all survive a deadly present.",
      popularity: 794.984,
      poster_path: "/rULWuutDcN5NvtiZi4FRPzRYWSh.jpg",
      release_date: "2024-02-14",
      title: "Madame Web",
      video: false,
      vote_average: 5.619,
      vote_count: 113,
    },
    {
      adult: false,
      backdrop_path: "/meyhnvssZOPPjud4F1CjOb4snET.jpg",
      genre_ids: [16, 28, 12, 35, 10751],
      id: 940551,
      original_language: "en",
      original_title: "Migration",
      overview:
        "After a migrating duck family alights on their pond with thrilling tales of far-flung places, the Mallard family embarks on a family road trip, from New England, to New York City, to tropical Jamaica.",
      popularity: 769.948,
      poster_path: "/ldfCF9RhR40mppkzmftxapaHeTo.jpg",
      release_date: "2023-12-06",
      title: "Migration",
      video: false,
      vote_average: 7.632,
      vote_count: 654,
    },
    {
      adult: false,
      backdrop_path: "/4MCKNAc6AbWjEsM2h9Xc29owo4z.jpg",
      genre_ids: [28, 53, 18],
      id: 866398,
      original_language: "en",
      original_title: "The Beekeeper",
      overview:
        "One man’s campaign for vengeance takes on national stakes after he is revealed to be a former operative of a powerful and clandestine organization known as Beekeepers.",
      popularity: 765.421,
      poster_path: "/A7EByudX0eOzlkQ2FIbogzyazm2.jpg",
      release_date: "2024-01-10",
      title: "The Beekeeper",
      video: false,
      vote_average: 7.3,
      vote_count: 1144,
    },
    {
      adult: false,
      backdrop_path: "/cnqwv5Uz3UW5f086IWbQKr3ksJr.jpg",
      genre_ids: [28, 12, 14],
      id: 572802,
      original_language: "en",
      original_title: "Aquaman and the Lost Kingdom",
      overview:
        "Black Manta seeks revenge on Aquaman for his father's death. Wielding the Black Trident's power, he becomes a formidable foe. To defend Atlantis, Aquaman forges an alliance with his imprisoned brother. They must protect the kingdom.",
      popularity: 744.491,
      poster_path: "/7lTnXOy0iNtBAdRP3TZvaKJ77F6.jpg",
      release_date: "2023-12-20",
      title: "Aquaman and the Lost Kingdom",
      video: false,
      vote_average: 6.907,
      vote_count: 1678,
    },
    {
      adult: false,
      backdrop_path: "/ehumsuIBbgAe1hg343oszCLrAfI.jpg",
      genre_ids: [16, 10751, 14, 12],
      id: 1022796,
      original_language: "en",
      original_title: "Wish",
      overview:
        "Asha, a sharp-witted idealist, makes a wish so powerful that it is answered by a cosmic force – a little ball of boundless energy called Star. Together, Asha and Star confront a most formidable foe - the ruler of Rosas, King Magnifico - to save her community and prove that when the will of one courageous human connects with the magic of the stars, wondrous things can happen.",
      popularity: 716.303,
      poster_path: "/nesuSdJakNkf0zs7OfoasB6Clxf.jpg",
      release_date: "2023-11-13",
      title: "Wish",
      video: false,
      vote_average: 6.649,
      vote_count: 735,
    },
    {
      adult: false,
      backdrop_path: "/8GnWDLn2AhnmkQ7hlQ9NJUYobSS.jpg",
      genre_ids: [18, 878, 28],
      id: 695721,
      original_language: "en",
      original_title: "The Hunger Games: The Ballad of Songbirds & Snakes",
      overview:
        "64 years before he becomes the tyrannical president of Panem, Coriolanus Snow sees a chance for a change in fortunes when he mentors Lucy Gray Baird, the female tribute from District 12.",
      popularity: 671.696,
      poster_path: "/mBaXZ95R2OxueZhvQbcEWy2DqyO.jpg",
      release_date: "2023-11-15",
      title: "The Hunger Games: The Ballad of Songbirds & Snakes",
      video: false,
      vote_average: 7.205,
      vote_count: 1754,
    },
    {
      adult: false,
      backdrop_path: "/gg4zZoTggZmpAQ32qIrP5dtnkEZ.jpg",
      genre_ids: [28, 80, 53],
      id: 891699,
      original_language: "en",
      original_title: "Silent Night",
      overview:
        "A tormented father witnesses his young son die when caught in a gang's crossfire on Christmas Eve. While recovering from a wound that costs him his voice, he makes vengeance his life's mission and embarks on a punishing training regimen in order to avenge his son's death.",
      popularity: 670.978,
      poster_path: "/nJCP1ZNTPKlZ7S0Kv3gbmuraAT4.jpg",
      release_date: "2023-11-30",
      title: "Silent Night",
      video: false,
      vote_average: 6.326,
      vote_count: 469,
    },
    {
      adult: false,
      backdrop_path: "/sRLC052ieEzkQs9dEtPMfFxYkej.jpg",
      genre_ids: [878, 18, 28],
      id: 848326,
      original_language: "en",
      original_title: "Rebel Moon - Part One: A Child of Fire",
      overview:
        "When a peaceful colony on the edge of the galaxy finds itself threatened by the armies of the tyrannical Regent Balisarius, they dispatch Kora, a young woman with a mysterious past, to seek out warriors from neighboring planets to help them take a stand.",
      popularity: 661.866,
      poster_path: "/ui4DrH1cKk2vkHshcUcGt2lKxCm.jpg",
      release_date: "2023-12-15",
      title: "Rebel Moon - Part One: A Child of Fire",
      video: false,
      vote_average: 6.4,
      vote_count: 1441,
    },
    {
      adult: false,
      backdrop_path: "/nTPFkLUARmo1bYHfkfdNpRKgEOs.jpg",
      genre_ids: [35, 10749],
      id: 1072790,
      original_language: "en",
      original_title: "Anyone But You",
      overview:
        "After an amazing first date, Bea and Ben’s fiery attraction turns ice cold — until they find themselves unexpectedly reunited at a destination wedding in Australia. So they do what any two mature adults would do: pretend to be a couple.",
      popularity: 636.132,
      poster_path: "/yRt7MGBElkLQOYRvLTT1b3B1rcp.jpg",
      release_date: "2023-12-21",
      title: "Anyone But You",
      video: false,
      vote_average: 6.937,
      vote_count: 429,
    },
    {
      adult: false,
      backdrop_path: "/zLj0peaxy5y2SlC6wNIQ4V0pfqg.jpg",
      genre_ids: [16, 10751, 35, 14],
      id: 1139829,
      original_language: "en",
      original_title: "Orion and the Dark",
      overview:
        "A boy with an active imagination faces his fears on an unforgettable journey through the night with his new friend: a giant, smiling creature named Dark.",
      popularity: 622.036,
      poster_path: "/N2ddXhBX5kqwA9flbm2yqFtX6b.jpg",
      release_date: "2024-02-02",
      title: "Orion and the Dark",
      video: false,
      vote_average: 6.699,
      vote_count: 214,
    },
    {
      adult: false,
      backdrop_path: "/yl2GfeCaPoxChcGyM5p7vYp1CKS.jpg",
      genre_ids: [28, 35, 10749],
      id: 848187,
      original_language: "en",
      original_title: "Role Play",
      overview:
        "Emma has a wonderful husband and two kids in the suburbs of New Jersey – she also has a secret life as an assassin for hire – a secret that her husband David discovers when the couple decide to spice up their marriage with a little role play.",
      popularity: 621.92,
      poster_path: "/7MhXiTmTl16LwXNPbWCmqxj7UxH.jpg",
      release_date: "2024-01-04",
      title: "Role Play",
      video: false,
      vote_average: 6.016,
      vote_count: 346,
    },
    {
      adult: false,
      backdrop_path: "/47SVqaO02doJ06tOmrjiWDkwU3T.jpg",
      genre_ids: [28, 53],
      id: 927107,
      original_language: "en",
      original_title: "The Bricklayer",
      overview:
        "Someone is blackmailing the CIA by assassinating foreign journalists and making it look like the agency is responsible. As the world begins to unite against the U.S., the CIA must lure its most brilliant – and rebellious – operative out of retirement, forcing him to confront his checkered past while unraveling an international conspiracy.",
      popularity: 578.84,
      poster_path: "/36pYugctLa70NmwMEgXTR1G31Kq.jpg",
      release_date: "2023-12-14",
      title: "The Bricklayer",
      video: false,
      vote_average: 6.115,
      vote_count: 113,
    },
  ],
  total_pages: 42500,
  total_results: 849996,
};
