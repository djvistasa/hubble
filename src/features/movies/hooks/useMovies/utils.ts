import { useMoviesStore } from "@features/movies/store";
import { IMovieResponse } from "@features/movies/types";
import { makeApiRequest } from "@utils";
import { SetURLSearchParams } from "react-router-dom";

const getPaginatedMoviesData = async (
  page: number,
  setSearchParams: SetURLSearchParams
) => {
  const { error, result, ok } = await makeApiRequest(
    `/discover/movie?include_adult=false&include_video=false&language=en-US&page=${page}&sort_by=popularity.desc`
  );

  if (error) {
    // handle error
    return [];
  }

  if (result && ok) {
    const { results, total_pages } = result as IMovieResponse;

    setSearchParams({
      total_pages: total_pages.toString(),
    });

    useMoviesStore.setState({ movies: results });
    useMoviesStore.setState({ filteredMovies: results });
  }
};

export { getPaginatedMoviesData };
