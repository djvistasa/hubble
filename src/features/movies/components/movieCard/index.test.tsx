import useTestComponentWithTheme from "@hooks/useTestComponentWithTheme";
import { describe, expect, it, vi } from "vitest";

import { moviesResp } from "@features/movies/hooks/useMovies/mocks";
import MovieCard from ".";

describe("MovieCard", () => {
  const mountWithTheme = useTestComponentWithTheme();
  const mockOnClick = vi.fn();
  const mockMovie = moviesResp.results[0];

  it("should render correctly", () => {
    const component = mountWithTheme(
      <MovieCard movie={mockMovie} onMovieClick={mockOnClick} />
    );
    expect(component).toMatchSnapshot();
  });

  it("should call onMovieClick when clicked", () => {
    const { getByTestId } = mountWithTheme(
      <MovieCard
        movie={mockMovie}
        onMovieClick={mockOnClick}
        testID="movie-card"
      />
    );
    const movieCard = getByTestId("movie-card");
    movieCard.click();
    expect(mockOnClick).toHaveBeenCalled();
  });
});
