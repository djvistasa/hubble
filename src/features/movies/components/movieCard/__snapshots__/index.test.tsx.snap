// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MovieCard > should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <a
        class="styles__StyledMovieCard-sc-vuoc3k-0 gZnIKa"
        data-testid="movie-card-Badland Hunters"
      >
        <div
          class="styles__StyledMask-sc-vuoc3k-2 LYXKU mask"
        />
        <p
          class="styles__StyledMovieInfoWrapper-sc-vuoc3k-1 hOBvPr"
        >
          Badland Hunters
        </p>
      </a>
    </div>
  </body>,
  "container": <div>
    <a
      class="styles__StyledMovieCard-sc-vuoc3k-0 gZnIKa"
      data-testid="movie-card-Badland Hunters"
    >
      <div
        class="styles__StyledMask-sc-vuoc3k-2 LYXKU mask"
      />
      <p
        class="styles__StyledMovieInfoWrapper-sc-vuoc3k-1 hOBvPr"
      >
        Badland Hunters
      </p>
    </a>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
